src/Microsoft.DotNet.Interactive/KernelExtensions.cs

```diff
@@ -211,7 +213,194 @@ public static T UseLogMagicCommand<T>(this T kernel)
+                    );
+                        var v = GetValue(fromKernel, valueProduced.Name, mimeType).GetAwaiter().GetResult();
+                        formattedValue = v.FormattedValue;
+                    }
+                    SendValue(kernel, referenceValue, formattedValue, valueName)
+                        .GetAwaiter().GetResult();
+                }
+                else
+                {
+                    var interpolatedValue = cmdLineContext.ParseResult.GetValueForOption(fromValueOption);
+                    SendValue(kernel, interpolatedValue,null, valueName)
+                        .GetAwaiter().GetResult();
+                }
+            }
+            else
+            {
+                context.Fail(context.Command, new CommandNotSupportedException(typeof(SendValue), kernel));
+            }
+        }
+    }
+
     public static T UseValueSharing<T>(this T kernel) where T : Kernel
+    {
+        ConfigureAndAddShareMagicCommand(kernel);
+        ConfigureAndAddSetMagicCommand(kernel);
+        return kernel;
+    }
+
+    private static void ConfigureAndAddSetMagicCommand<T>(T kernel) where T : Kernel
+    {
+        var fromResultOption = new Option<bool>(
+            "--from-result",
+            description: "Captures the execution result.");
+
+
+        var fromValueOption = new Option<string>(
+            "--from-value",
+            description: "Specifies a value to be stored directly. Specifying @input:value allows you to prompt the user for this value.",
+            parseArgument: result =>
+            {
+                if (SetErrorIfAlsoUsed(fromResultOption, result))
+                {
+                    return null;
+                }
+
+                return result.Tokens.Single().Value;
+            });
+
+        var mimeTypeOption = new Option<string>("--mime-type", "Share the value as a string formatted to the specified MIME type.")
+            .AddCompletions(
+                JsonFormatter.MimeType,
+                HtmlFormatter.MimeType,
+                PlainTextFormatter.MimeType);
+
+        var nameOption = new Option<string>(
+            "--name",
+            description: "This is the name used to declare and set the value in the kernel.",
+            parseArgument: result =>
+            {
+                if (SetErrorIfNoneAreUsed(result, fromResultOption, fromValueOption))
+                {
+                    return null;
+                }
+
+                return result.Tokens.Single().Value;
+            }
+        )
+        {
+            IsRequired = true
+        };
+
+        var set = new Command("#!set")
+        {
+            nameOption,
+            fromResultOption,
+            fromValueOption,
+            mimeTypeOption
+        };
+
+        set.SetHandler(cmdLineContext =>
+        {
+            HandleSetMagicCommand(kernel, cmdLineContext, fromResultOption, nameOption, fromValueOption, mimeTypeOption);
```

Comment:
Since this is also being used by `#!share`, a more generalized name for `HandleSetMagicCommand` would be clearer.